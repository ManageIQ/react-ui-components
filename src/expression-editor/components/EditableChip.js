import React from 'react';
import PropTypes from 'prop-types';
import Chip from './Chip';

import AutocompleteTextInput from './AutocompleteTextInput';
import ExpressionEditorPropTypes from './ExpressionEditorPropTypes';

class EditableChip extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      filteredOptions: props.options,
      filterString: this.props.label,
      aliasMode: false,
      // filterString: '',
    };
    // console.log('editable chip constructor', props);
  }

  static getDerivedStateFromProps(props, state) {
    // console.log('editable chipget derived state', props, state);
    return { filteredOptions: props.options.filter(props.filterOption(state.filterString)) };
  }

  onClick = () => {
    this.props.onClick(this.props.index);
  };

  onDoubleClick = () => {
    this.setState({ filterString: this.props.label });
    this.props.onDoubleClick(this.props.selected, this.props.index);
  };

  onSubmit = (selected) => {
    // console.log('EditableChip submit', selected, this.props.item);
    console.log('MODE:', this.state.aliasMode);
    if (this.state.aliasMode) {
      this.setAlias(selected);
      this.setState({aliasMode: false});
    } else {
      const newSelected = selected;
      if (selected.type === 'userinput') {
        // console.log('SUMBIT:',selected);
        newSelected.parent = this.props.item.parent;
      }
      // let{ label, id, item } = this.props;
      this.setState({ filterString: '' });
      this.props.onSubmit(newSelected, this.props.item);
    }
  }

  onDelete = () => {
    // console.log('on DELETEE');
    this.props.onDelete(this.props.item, this.props.index);
  }

  onFocus = () => {
    // console.log('on FOCUSSSSSSSS');
    this.props.onFocus(this.props.item, this.props.index);
  }

  onBlur = () => {
    // console.log('on bluuuuuuuuuuuuuuuuuur');
    this.props.onBlur(this.props.item, this.props.index);
  }

  onUserInput = (input) => {
  // console.log('EditableChip UserInput', input);
    this.setState({ filterString: input });
  }

  onKeyDown = (key) => {
    this.props.onKeyDown(key, this.props.index, this.props.selected);
  }

  setAliasMode = (bool) => {
    this.setState({aliasMode: bool});
  }

  setAlias = (alias) => {
    console.log('ALIAS',alias, this.props.index );
    this.props.setAlias(alias, this.props.index)
  }

  render() {
    // console.log('EditableChip props:', this.props.label);
    return (
      (this.props.isEditing &&
        <AutocompleteTextInput
          onSubmit={this.onSubmit}
          onChange={this.onUserInput}
          onKeyDown={this.onKeyDown}
          options={this.state.filteredOptions}
          value={this.state.filterString}
          denyUserInput={(this.props.options.length > 0)}
          // setAlias={this.props.setAlias}
          setAliasMode={this.setAliasMode}
          aliasMode={this.state.aliasMode}
          isLastElement={!!this.props.isLastElement}
          deleteExpression={this.props.deleteExpression}
          blurAllChips={this.props.blurAllChips}
          // registerInput={this.props.registerInput}
          // unregisterInput={this.props.unregisterInput}
          inputRef={this.props.inputRef}
          index={this.props.index}
          next={this.props.item.next}
        />) ||
        <Chip
          onDelete={this.onDelete}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onKeyDown={this.onKeyDown}
          registerChip={this.props.registerChip}
          unregisterChip={this.props.unregisterChip}
          isAliasSet={this.props.isAliasSet}
          setAliasMode={this.setAliasMode}
          setAlias={this.setAlias}
          isFocused={this.props.isFocused}
          index={this.props.index}
          onClick={this.onClick}
          onDoubleClick={this.onDoubleClick}
          label={this.props.label}
          type={this.props.selected.type}
          chipRef={this.props.chipRef}
        />
    );
  }
}

EditableChip.propTypes = {
  onClick: PropTypes.func.isRequired,
  onDoubleClick: PropTypes.func.isRequired,
  onSubmit: PropTypes.func,
  onKeyDown: PropTypes.func,
  onDelete: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  setAlias: PropTypes.func,
  deleteExpression: PropTypes.func,
  registerChip: PropTypes.func,
  unregisterChip: PropTypes.func,
  label: PropTypes.string.isRequired,
  item: PropTypes.object,
  index: PropTypes.number.isRequired,
  isAliasSet: PropTypes.bool,
  isFocused: PropTypes.bool.isRequired,
  isEditing: PropTypes.bool,
  options: PropTypes.arrayOf(ExpressionEditorPropTypes.term).isRequired,
  selected: ExpressionEditorPropTypes.option,
  filterOption: PropTypes.func,
  inputRef: PropTypes.object,
  chipRef: PropTypes.object,
};

EditableChip.defaultProps = {
  item: {},
  selected: {},
  isEditing: true,
  filterOption: value => option => option.label.toLowerCase().includes(value.toLowerCase()),
};


export default EditableChip;
