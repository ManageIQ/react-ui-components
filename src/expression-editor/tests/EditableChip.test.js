import React from 'react';
import EditableChip from '../components/EditableChip';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

const chipProps = {
  key: 1,
  label: 'label',
  isAliasSet: false,
  isEditing: false,
  isFocused: false,
  onClick: jest.fn(),
  onDoubleClick: jest.fn(),
  onSubmit: jest.fn(),
  onDelete: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
  setAlias: jest.fn(),
  blurAllChips: jest.fn(),
  onKeyDown: jest.fn(),
  chipRef: React.createRef(),
  options: [],
  selected: {id: 111, label: 'first', type: 'operator', next: []},
  item: {id: 111, label: 'first', type: 'operator', next: []},
  index: 0,
};

describe('EditableChip Component', () => {

  it('Onclick calls callback', () => {
    let onClick = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} onClick={onClick}/>
    );

    // let spy = jest.spyOn(wrapper.instance(), 'setAliasMode');
    wrapper.find('span.pf-c-label').simulate('click');
    expect(onClick).toBeCalledWith(0);
  });

  it('onDoubleClick calls callback', () => {
    let onDoubleClick = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} onDoubleClick={onDoubleClick}/>
    );

    wrapper.find('span.pf-c-label').simulate('dblclick');
    expect(onDoubleClick).toBeCalledWith({"id": 111, "label": "first", "next": [], "type": "operator"}, 0);
  });

  it('onFocus calls callback', () => {
    let onFocus = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} onFocus={onFocus}/>
    );

    wrapper.find('span.pf-c-label').simulate('focus');
    expect(onFocus).toBeCalledWith({"id": 111, "label": "first", "next": [], "type": "operator"}, 0);
  });

  it('onBlur calls callback', () => {
    let onBlur = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} onBlur={onBlur}/>
    );

    wrapper.find('span.pf-c-label').simulate('blur');
    expect(onBlur).toBeCalledWith({"id": 111, "label": "first", "next": [], "type": "operator"}, 0);
  });

  it('onDelete calls callback', () => {
    let onDelete = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} onDelete={onDelete} isFocused/>
    );

    wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
    wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
    wrapper.find('span.focusedChip').simulate('keydown', {key: 'Enter', keyCode: 13});
    expect(onDelete).toBeCalledWith({"id": 111, "label": "first", "next": [], "type": "operator"}, 0);
  });

  it('onSubmit calls callback', () => {
    let onSubmit = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} isEditing onSubmit={onSubmit}/>
    );

    wrapper.find('input').simulate('keydown', {key: 'Enter', keyCode: 13});

    expect(onSubmit).toBeCalledWith(
      {"id": "label", "label": "label", "next": [], "parent": undefined, "type": "userinput"},
      {"id": 111, "label": "first", "next": [], "type": "operator"}
    );
  });

  it('onSubmit calls callback with setAlias=true', () => {

    let setAlias = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} isEditing setAlias={setAlias}/>
    );

    wrapper.setState({aliasMode: true});
    wrapper.find('input').simulate('keydown', {key: 'Enter', keyCode: 13});
    expect(wrapper.state().aliasMode).toBe(false);
    expect(setAlias).toBeCalledWith('label', 0);
  });

  it('onChange calls callback', () => {
    let onChange = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} isEditing onChange={onChange} label='ahoj'/>
    );

    wrapper.find('input').simulate('change');
    expect(wrapper.state().filterString).toBe('ahoj');
  });

  it('setAliasMode sets state', () => {
    let setAlias = jest.fn();
    const wrapper = mount(
      <EditableChip {...chipProps} setAlias={setAlias}/>
    );

    let spy = jest.spyOn(wrapper.instance(), 'setAliasMode');
    let ret = wrapper.instance().setAliasMode('true');
    expect(wrapper.state().aliasMode).toBeTruthy();
  });



  it('Matches the snapshot with disabled edit', () => {
    const component = shallow(
      <EditableChip {...chipProps} />
    );

    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  it('Matches the snapshot with enabled edit', () => {
    const component = shallow(
      <EditableChip {...chipProps} isEditing={true} inputRef={React.createRef()} />
    );

    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
