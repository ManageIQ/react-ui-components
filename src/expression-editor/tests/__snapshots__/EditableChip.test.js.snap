// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableChip Component Matches the snapshot with disabled edit 1`] = `
<Chip
  chipRef={
    Object {
      "current": <span
        class=""
        id="pf-toggle-id-16"
        parentref="[object HTMLDivElement]"
        tabindex="0"
      >
        <span
          class="pf-c-label"
        >
          label
        </span>
      </span>,
    }
  }
  index={0}
  isAliasSet={false}
  isFocused={false}
  label="label"
  onBlur={[Function]}
  onClick={[Function]}
  onDelete={[Function]}
  onDoubleClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  setAlias={[Function]}
  setAliasMode={[Function]}
  type="operator"
/>
`;

exports[`EditableChip Component Matches the snapshot with enabled edit 1`] = `
<AutocompleteTextInput
  aliasMode={false}
  blurAllChips={[MockFunction]}
  denyUserInput={false}
  index={0}
  inputRef={
    Object {
      "current": null,
    }
  }
  isLastElement={false}
  matchingFunction={[Function]}
  next={Array []}
  onChange={[Function]}
  onKeyDown={[Function]}
  onSubmit={[Function]}
  options={Array []}
  setAliasMode={[Function]}
  value="label"
/>
`;
